client/comms/wsmanager.go

```diff
@@ -0,0 +1,200 @@
+package comms
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/url"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/davecgh/go-spew/spew"
+	"github.com/decred/dcrdex/server/comms/msgjson"
+	"github.com/gorilla/websocket"
+)
+
+const (
+	// pingWait is the maximum time in seconds to wait for a ping from
+	// the server. The client can be reconnected after this threshold since
+	// the websocket connection to the server is assumed to broken.
+	pingWait = 60
+)
+
+type ws interface {
+	ReadMessage() (int, []byte, error)
+	WriteJSON(v interface{}) error
+	SetPongHandler(handler func(data string) error)
+	SetReadDeadline(t time.Time) error
+	Close() error
+}
+
+type WsManager struct {
+	rID      uint64 // update atomically.
+	lastPong int64  // update atomically.
+	host     string
+	ws       ws
+	readCh   chan *msgjson.Message
+	sendCh   chan *msgjson.Message
+	req      map[uint64]string
+	reqMtx   sync.RWMutex
+	wg       sync.WaitGroup
+}
+
+// NewWsManager creates a client websocket manager.
+func NewWsManager(host string, ws ws) *WsManager {
+	return &WsManager{
+		host:   host,
+		ws:     ws,
+		readCh: make(chan *msgjson.Message),
+		sendCh: make(chan *msgjson.Message),
+		req:    make(map[uint64]string),
+	}
+}
+
+// logRoute logs a request route as an id/route pair.
+func (mgr *WsManager) logRoute(id uint64, route string) {
+	mgr.reqMtx.Lock()
+	mgr.req[id] = route
+	mgr.reqMtx.Unlock()
+}
+
+// fetchRoute fetches the route of the provided request id.
+func (mgr *WsManager) fetchRoute(id uint64) (string, error) {
+	mgr.reqMtx.RLock()
+	route := mgr.req[id]
+	mgr.reqMtx.RUnlock()
+
+	if route == "" {
+		return "", fmt.Errorf("no route found for id #%v", id)
+	}
+
+	return route, nil
+}
```

Comment:
These two functions aren't used, but we will probably want something like this. We'll have to figure out the best way to do it. Are we going to expect the user to maintain a mapping from the id to a response handler, or do we want this package to handle it, similar to what `comms` does?